File: problems/problems_437/solution.ts
import {TreeNode,JSONArrayToTreeNode} from "../../typescript/models/treenode";

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number {
    const dfs = (node: TreeNode | null, counter: Map<number, number>, sum: number): number => {
		if (node === null) {
			return 0;
		}
		let result: number = 0;
		sum += node.val;
		result += counter.get(sum - targetSum) || 0;
		counter.set(sum, (counter.get(sum) || 0) + 1);
		result += dfs(node.left, counter, sum);
		result += dfs(node.right, counter, sum);
		counter.set(sum, (counter.get(sum) || 0) - 1);
		return result;
	}
	return dfs(root, new Map<number, number>().set(0, 1), 0);
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const root: TreeNode | null = JSONArrayToTreeNode(JSON.parse(inputValues[0]));
	const targetSum: number = JSON.parse(inputValues[1]);
	return pathSum(root, targetSum);
}

File: problems/problems_437/solution.rs
use serde_json::{json, Value};
use library::lib::tree_node::{TreeNode, array_to_tree};
pub struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;
impl Solution {
    pub fn path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> i32 {
		fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, counter: &mut HashMap<i64, i32>, target_sum: i32, current_sum: i64) -> i32 {
			if node.is_none() {
				return 0;
			}
			let node = node.as_ref().unwrap().borrow();
			let mut res = 0;
			let current_sum = current_sum + node.val as i64;
			res += counter.get(&(current_sum - target_sum as i64)).unwrap_or(&0);
			*counter.entry(current_sum).or_insert(0) += 1;
			res += dfs(&node.left, counter, target_sum, current_sum);
			res += dfs(&node.right, counter, target_sum, current_sum);
			*counter.get_mut(&current_sum).unwrap() -= 1;
			res
		}
		let mut counter = HashMap::new();
		counter.insert(0, 1);
		dfs(&root, &mut counter, target_sum, 0)
    }
}

#[cfg(feature = "solution_437")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let input_vec0: Vec<Option<i32>> = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	let root: Option<Rc<RefCell<TreeNode>>> = array_to_tree(&input_vec0);
	let target_sum: i32 = serde_json::from_str(&input_values[1]).expect("Failed to parse input");
	json!(Solution::path_sum(root, target_sum))
}
